src/libmongoc/src/mongoc/mongoc-topology.c

```diff
@@ -1436,45 +1434,53 @@ _mongoc_topology_pop_server_session (mongoc_topology_t *topology,
    int64_t timeout;
    mongoc_server_session_t *ss = NULL;
    mongoc_topology_description_t *td;
+   bool loadbalanced;
 
    ENTRY;
 
    bson_mutex_lock (&topology->mutex);
 
    td = &topology->description;
    timeout = td->session_timeout_minutes;
+   loadbalanced = td->type == MONGOC_TOPOLOGY_LOAD_BALANCED;
 
-   if (timeout == MONGOC_NO_SESSIONS) {
-      /* if needed, connect and check for session timeout again */
-      if (!mongoc_topology_description_has_data_node (td)) {
-         bson_mutex_unlock (&topology->mutex);
-         if (!mongoc_topology_select_server_id (
-                topology, MONGOC_SS_READ, NULL, error)) {
-            RETURN (NULL);
-         }
+   /* When the topology type is LoadBalanced, sessions are always supported. */
+   if (!loadbalanced) {
+      if (timeout == MONGOC_NO_SESSIONS) {
+         /* if needed, connect and check for session timeout again */
+         if (!mongoc_topology_description_has_data_node (td)) {
+            bson_mutex_unlock (&topology->mutex);
+            if (!mongoc_topology_select_server_id (
+                   topology, MONGOC_SS_READ, NULL, error)) {
+               RETURN (NULL);
+            }
 
-         bson_mutex_lock (&topology->mutex);
-         timeout = td->session_timeout_minutes;
-      }
+            bson_mutex_lock (&topology->mutex);
+            timeout = td->session_timeout_minutes;
+         }
 
-      if (timeout == MONGOC_NO_SESSIONS) {
-         bson_mutex_unlock (&topology->mutex);
-         bson_set_error (error,
-                         MONGOC_ERROR_CLIENT,
-                         MONGOC_ERROR_CLIENT_SESSION_FAILURE,
-                         "Server does not support sessions");
-         RETURN (NULL);
+         if (timeout == MONGOC_NO_SESSIONS) {
+            bson_mutex_unlock (&topology->mutex);
+            bson_set_error (error,
+                            MONGOC_ERROR_CLIENT,
+                            MONGOC_ERROR_CLIENT_SESSION_FAILURE,
+                            "Server does not support sessions");
+            RETURN (NULL);
+         }
       }
    }
 
    while (topology->session_pool) {
       ss = topology->session_pool;
       CDL_DELETE (topology->session_pool, ss);
-      if (_mongoc_server_session_timed_out (ss, timeout)) {
-         _mongoc_server_session_destroy (ss);
-         ss = NULL;
-      } else {
-         break;
+      /* Sessions do not expire when the topology type is load balanced. */
+      if (!loadbalanced) {
```

Comment:
To minimize the diff and conditional nesting, I'd suggest:

```c
if (!loadBalanced) {
   continue;
}
```

I assume you still want to clear the pool regardless of topology. It's just killing of sessions that you want to avoid.